src/xopen/__init__.py

```diff
@@ -650,6 +652,50 @@ def __init__(
         )
 
 
+class PipedXzReader(PipedCompressionReader):
+    """
+    Open a pipe to xz for reading an xz-compressed file.
+    """
+
+    # TODO: implement threaded decompression once XZ releases it
```

Comment:
Is this something that is planned from the xz side? It would be nicer to remove this comment and instead file an xopen issue to remind us.

Reply:
There is also `pixz` a parallel implementation of  `xz` . But it is less useful as xz already has threaded support.

From the XZ manual 

>   Threaded  decompression  hasn't  been  implemented yet.  It will
only work on files that contain multiple blocks with size infor‐
mation in block headers.  All files compressed in multi-threaded
mode meet  this  condition,  but  files  compressed  in  single-
threaded mode don't even if --block-size=size is used.

That means that adding a threads flag PipedXzReader is already valuable **now** as files that have been compressed in xz's threaded mode will decompress much faster when threaded mode is enabled for decompression.

Reply:
> Is this something that is planned from the xz side?

I noticed that it was in xz's `master` branch (which hasn't been released yet), so I didn't know if it would be useful to pass the `-T` flag to the decompression stage. See:

- https://git.tukaani.org/?p=xz.git;a=commitdiff;h=4cce3e27f529af33e0e7749a8cbcec59954946b5
- https://git.tukaani.org/?p=xz.git;a=commitdiff;h=6c6da57ae2aa962aabde6892442227063d87e88c

Reply:
Interesting. `pigz` also has "multithreaded" decompression, but Mark Adler stated that due to the format it was not really possible. So what he did is load off decompression, input processing and output processing to different threads, so that the decompression thread could use 100% of the CPU. 

The result is one thread that uses 100% and the whole process uses 130% (1.3 cores.) This 30% increase in CPU time leads to a 10% increase in throughput. 
For this reason xopen defaults to 1 thread for pigz whilst decompressing. 

It will be interesting to see how this dynamic works out for XZ. I get the feeling that when blocks have separate headers that speed will scale very well with number of threads. If not I expect more Pigz-like speed. If they manage to beat that expectation than it is pretty good indeed.